plugins/toolbarwatcher/plugin.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+(function () {
+	'use strict';
+
+	var PLUGIN_NAME = 'toolbarwatcher';
+
+	if (!CKEDITOR.plugins.get(PLUGIN_NAME)) {
+		var TOOLBAR_HIDDEN_LABELS_CLASS = 'cke_hide_toolbar_labels';
+
+		var stylesLoaded = false;
+
+		function debounce(fn, delay) {
+			return function debounced() {
+				var args = Array.prototype.slice.call(arguments);
+				cancelDebounced(debounced);
+				debounced.id = setTimeout(function () {
+					fn.apply(null, args);
+				}, delay);
+			};
+		}
+
+		function cancelDebounced(debounced) {
+			clearTimeout(debounced.id);
+		}
+
+		CKEDITOR.plugins.add(PLUGIN_NAME, {
+			afterInit(editor) {
+				this.editor = editor;
+				this.editor.on(
+					'instanceReady',
+					this.onInstanceReady.bind(this)
+				);
+			},
+
+			editor: null,
+
+			getToolbarsWidth() {
+				var containerElement = this.editor.container.$;
+
+				var toolbars = Array.from(
+					containerElement.querySelectorAll('.cke_toolbar')
+				);
+
+				return toolbars.reduce(function (acc, toolbar) {
+					return acc + toolbar.offsetWidth;
+				}, 0);
+			},
+
+			init(editor) {
+				if (!stylesLoaded) {
+					CKEDITOR.document.appendStyleSheet(
+						this.path + 'styles/' + PLUGIN_NAME + '.css'
+					);
+					stylesLoaded = true;
+				}
+			},
+
+			isToolbarOverflowing() {
+				var toolbarsContainer = this.editor.container.$.querySelector(
+					'.cke_top'
+				);
+
+				var toolbarsContainerWidth = Infinity;
+
+				if (toolbarsContainer) {
+					toolbarsContainerWidth =
+						parseInt(
+							getComputedStyle(toolbarsContainer).width,
+							10
+						) || 0;
+				}
+
+				return this.labeledToolbarsWidth >= toolbarsContainerWidth;
```

Comment:
(Subjective, but) the `Infinity` here is a bit distracting because the thing reads as "magical special value" when I think the actual intent is "if I have a container, do a calculation with it, otherwise don't". So this might be clearer:

```js
if (toolbarsContainer) {
    var width = parseInt(
        getComputedStyle(toolbarsContainer).width,
        10
    ) || 0;

    return this.labeledToolbarsWidth >= width;
}

return false;
```

Up to you though.

Reply:
![](https://media.giphy.com/media/HufOeXwDOInlK/giphy.gif)

Reply:
good point